From 8f528217e008d1c1357dc5a7ed9a852007ac1ea8 Mon Sep 17 00:00:00 2001
From: damoferenc <damo.ferenc@yahoo.com>
Date: Thu, 5 Aug 2021 16:27:44 +0300
Subject: [PATCH] drmP.h include deleted

---
 drivers/gpu/drm/xlnx/digilent_hdmi.c | 1 -
 1 file changed, 1 deletion(-)

diff --git a/drivers/gpu/drm/xlnx/digilent_hdmi.c b/drivers/gpu/drm/xlnx/digilent_hdmi.c
index 388b73de27bc..3df11a876b4c 100644
--- a/drivers/gpu/drm/xlnx/digilent_hdmi.c
+++ b/drivers/gpu/drm/xlnx/digilent_hdmi.c
@@ -8,7 +8,6 @@
  */
 
 
-#include <drm/drmP.h>
 #include <drm/drm_atomic_helper.h>
 #include <drm/drm_crtc.h>
 #include <drm/drm_fourcc.h>
-- 
2.17.1

